File: backend/utils/cdn.js
import crypto from 'crypto';
import { DeleteObjectCommand, HeadObjectCommand, ListObjectsV2Command, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

const bucket = process.env.CDN_BUCKET_NAME;
const publicUrl = process.env.CDN_PUBLIC_URL?.replace(/\/$/, '');

if (!process.env.CDN_ENDPOINT || !process.env.CDN_ACCESS_KEY_ID || !process.env.CDN_SECRET_ACCESS_KEY || !bucket || !publicUrl) {
  throw new Error('CDN R2 configuration is incomplete');
}

const client = new S3Client({
  region: 'auto',
  endpoint: process.env.CDN_ENDPOINT,
  credentials: {
    accessKeyId: process.env.CDN_ACCESS_KEY_ID,
    secretAccessKey: process.env.CDN_SECRET_ACCESS_KEY,
  },
});

// Keep media namespaces stable so database URLs and cleanup remain predictable.
export const CDN_PREFIXES = {
  reels: 'reels',
  profilePictures: 'profile_pictures',
};

export const createObjectKey = (prefix, userId, fileName = '') => {
  const extension = fileName.includes('.') ? `.${fileName.split('.').pop().toLowerCase().replace(/[^a-z0-9]/g, '')}` : '';
  return `${prefix}/${userId}/${Date.now()}-${crypto.randomUUID()}${extension}`;
};

export const publicObjectUrl = (key) => `${publicUrl}/${key}`;

// Ownership checks prevent users from attaching or deleting another user's media.
export const isOwnedObjectUrl = (url, prefix, userId) =>
  typeof url === 'string' && url.startsWith(`${publicUrl}/${prefix}/${userId}/`);

export const presignUpload = async ({ key, contentType, expiresIn = 900 }) => {
  const command = new PutObjectCommand({ Bucket: bucket, Key: key, ContentType: contentType });
  return getSignedUrl(client, command, { expiresIn });
};

// Verify that the client completed the direct upload before writing its URL to the database.
export const headObject = (key) => client.send(new HeadObjectCommand({ Bucket: bucket, Key: key }));

export const deleteObject = (key) => client.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));

export const deletePrefix = async (prefix) => {
  let continuationToken;
  do {
    const listed = await client.send(new ListObjectsV2Command({ Bucket: bucket, Prefix: prefix, ContinuationToken: continuationToken }));
    await Promise.all((listed.Contents || []).map(({ Key }) => Key && deleteObject(Key)));
    continuationToken = listed.IsTruncated ? listed.NextContinuationToken : undefined;
  } while (continuationToken);
};

export const getBucketName = () => bucket;

File: reels-service/app/core/cdn.py
import asyncio
import re
import time
import uuid

import boto3
from botocore.config import Config as BotoConfig

from app.core.config import settings

# Port of backend/utils/cdn.js - both services share one R2 bucket, so key
# layout, ownership rules, and public URL shape must stay identical to what
# the Node backend already wrote to the database.

CDN_PREFIXES = {
    "reels": "reels",
    "profile_pictures": "profile_pictures",
}

_client = boto3.client(
    "s3",
    region_name="auto",
    endpoint_url=settings.CDN_ENDPOINT,
    aws_access_key_id=settings.CDN_ACCESS_KEY_ID,
    aws_secret_access_key=settings.CDN_SECRET_ACCESS_KEY,
    # Virtual-hosted-style ("<bucket>.<account>.r2.cloudflarestorage.com"),
    # matching the Node backend's AWS SDK v3 client (no forcePathStyle set,
    # which defaults to virtual-hosted there) - kept identical so presigned
    # URLs from either service hit R2 the same way.
    config=BotoConfig(signature_version="s3v4", s3={"addressing_style": "virtual"}),
)


def create_object_key(prefix: str, user_id: int, file_name: str = "") -> str:
    extension = ""
    if "." in file_name:
        raw_ext = file_name.rsplit(".", 1)[-1].lower()
        cleaned_ext = re.sub(r"[^a-z0-9]", "", raw_ext)
        extension = f".{cleaned_ext}" if cleaned_ext else ""
    return f"{prefix}/{user_id}/{int(time.time() * 1000)}-{uuid.uuid4()}{extension}"


def public_object_url(key: str) -> str:
    return f"{settings.CDN_PUBLIC_URL.rstrip('/')}/{key}"


# Ownership checks prevent users from attaching or deleting another user's media.
def is_owned_key(key: str, prefix: str, user_id: int) -> bool:
    return isinstance(key, str) and key.startswith(f"{prefix}/{user_id}/") and ".." not in key


def presign_upload(key: str, content_type: str, expires_in: int = 900) -> str:
    return _client.generate_presigned_url(
        "put_object",
        Params={"Bucket": settings.CDN_BUCKET_NAME, "Key": key, "ContentType": content_type},
        ExpiresIn=expires_in,
    )


# Verify that the client completed the direct upload before writing its URL
# to the database. Runs the blocking boto3 call off the event loop.
async def head_object(key: str) -> None:
    await asyncio.to_thread(_client.head_object, Bucket=settings.CDN_BUCKET_NAME, Key=key)
